### PR TITLE
fix cdn race condition

### PR DESCRIPTION
**Pull request by silesky**
- I'm still evaluating if it's safe to remove the `setGlobalCDNUrl` call from `brower-umd.ts`  and `standalone.ts` (if anyone has any thoughts / I'm interested!).

### CODE REVIEW

File: example/pages/index.tsx [hunk 1]
```diff
@@ -31,13 +37,57 @@ const jsontheme = {
...
+  }
+  return [storedValue, setValue] as const
+}
+
 export default function Home(): React.ReactElement {
   const [analytics, setAnalytics] = useState<Analytics | undefined>(undefined)
-  const [settings, setSettings] = useState<AnalyticsSettings | undefined>(
-    undefined
-  )
+  const [writeKey, setWriteKey] = useLocalStorage<string>('__dev_writeKey', '')
```
Comment (silesky) [inline example/pages/index.tsx:82]: This is a somewhat unrelated change that I'm sneaking in -- I wanted to be able to have the write key input persist between page refreshes.

File: src/plugins/ajs-destination/loader.ts [hunk 2]
```diff
@@ -42,6 +39,7 @@ export async function loadIntegration(
 ): Promise<LegacyIntegration> {
   const pathName = normalizeName(name)
   const obfuscatedPathName = obfuscatePathName(pathName, obfuscate)
+  const path = getNextIntegrationsURL()
```
Comment (silesky) [inline src/plugins/ajs-destination/loader.ts:42]: calling this method outside of the global scope make them easier to reason about, especially since it's a wrapper around a global, we don't need to worry about stale references.

File: example/pages/index.tsx [hunk 3]
```diff
@@ -31,13 +37,57 @@ const jsontheme = {
...
+  const [settings, setSettings] = useState<
+    AnalyticsBrowserSettings | undefined
+  >({
+    writeKey,
+    cdnURL: 'https://cdn.segment.com',
+  })
```
Comment (yordis) [inline example/pages/index.tsx:88]:
```suggestion
  const [writeKey, setWriteKey] = useLocalStorage<string>('__dev_writeKey', '')
  const [settings, setSettings] = useState<
    AnalyticsBrowserSettings | undefined
  >({
    writeKey: process.env.NEXT_PUBLIC_SEGMENT_WRITE_KEY,
    cdnURL: process.env.NEXT_PUBLIC_SEGMENT_CDN_URL,
  })
```

I would not like to teach people to do this ... it will cause issues like caching the wrong value or somebody blocking the `localStorage` ... you named it.

You do you, but I rather teach people to put such information in the `env` file since that is probably where people should be injecting such thing, and it will fix your issue

File: src/lib/parse-cdn.ts [hunk 4]
```diff
@@ -17,14 +17,17 @@ const getCDNUrlFromScriptTag = (): string | undefined => {
   return cdn
 }
 
+let _globalCDN: string | undefined // set globalCDN as in-memory singleton
```
Comment (yordis) [inline src/lib/parse-cdn.ts:20]: I LOVE IT

File: src/lib/parse-cdn.ts [hunk 5]
```diff
@@ -45,6 +48,12 @@ export const getCDN = (): string => {
   }
 }
 
+export const getNextIntegrationsURL = () => {
+  const cdn = getCDN()
+  const path = `${cdn}/next-integrations`
```
Comment (yordis) [inline src/lib/parse-cdn.ts:53]:
```suggestion
  return `${cdn}/next-integrations`
```

Interesting, I am curious to learn what made you decide to use anchor variables style rather than inlining the return 🤔 Nothing here btw, I am just curious of the habit.

Comment (silesky) [reply src/lib/parse-cdn.ts:53]: That’s a good question!

In part, it was copypasta from the refactor. 

On the broad topic, I’m pretty inconsistent, but I do feel that raw return make debugging harder for “complex” expressions like “return pipe(doFoo, doBar)”, since setting a breakpoint to capture a function’s return value can be tricky if not impossible for many debuggers, so you often have to go to the caller to set the breakpoint. I have no hard and fast rules — I think intermediate variables are one of those things that I don’t attach hard and fast rules to, it’s a tool to communicate developer intent.

Here, I think we could definitely do without the return, as it’s quite simple.

Comment (yordis) [reply src/lib/parse-cdn.ts]: > On the broad topic, I’m pretty inconsistent, but I do feel that raw return make debugging harder for “complex” expressions like “return pipe(doFoo, doBar)”,

Right right, I can see that some debuggers allow you to step into the return statement itself, while the part is to step out and see the value there rather than caring at the function call level ... 

I see I see ... just trying to learn about you in general.

Thank you for responding thou 💜

File: example/pages/index.tsx (hunk 3 above)
Comment (silesky) [reply example/pages/index.tsx:88]: - I didn't even consider that people might want to override cdnURL, but that's a good idea!
- on the subject of writeKey -- on a regular application, I agree. However, this is a playground for debugging (run `make dev` to see) -- so there's a writeKey input field that allows people to swap out writeKeys at runtime.

**Comment by pooyaj** [comment]
@silesky I think the purpose of the `setGlobalCDNUrl` in the standalone etc. is to avoid consistently hitting the `getCDNUrlFromScriptTag` when calling `getCDN`, and only using `getCDNUrlFromScriptTag` once ( because it is a relatively costly DOM scan ).

File: example/pages/index.tsx (hunk 3 above)
Comment (yordis) [reply example/pages/index.tsx]: > I didn't even consider that people might want to override cdnURL, but that's a good idea!

Well, we have to do a proxy to Segment to bypass AdBlocker ... so probably most companies that know a bit about infra would do that most of the time 😄

Comment (yordis) [reply example/pages/index.tsx]: > run make dev if you haven't yet to see what I mean!

I see I see, would call the directory "playground" because people are getting used to the idea of using `examples` directory to learn how to get up to speed with this ... which is my actually worry (not you, but others reading code out of context)

Comment (silesky) [reply example/pages/index.tsx]: > > I didn't even consider that people might want to override cdnURL, but that's a good idea!
> 
> Well, we have to do a proxy to Segment to bypass AdBlocker ... so probably most companies that know a bit about infra would do that most of the time 😄

yep, I just meant the few of us who use the playground — I’d just seen it used to demo segment and develop locally directly against segment itself. Anyway, now it’s overridavle!

Comment (silesky) [reply example/pages/index.tsx]: > > run make dev if you haven't yet to see what I mean!
> 
> I see I see, would call the directory "playground" because people are getting used to the idea of using `examples` directory to learn how to get up to speed with this ... which is my actually worry (not you, but others reading code out of context)

I think it’s meant to serve a duel role, but that’s a good point — it’s not typically what you think of as an ‘example’ in an examples repo!

**Comment by silesky** [comment]
> @silesky I think the purpose of the `setGlobalCDNUrl` in the standalone etc. is to avoid consistently hitting the `getCDNUrlFromScriptTag` when calling `getCDN`, and only using `getCDNUrlFromScriptTag` once ( because it is a relatively costly DOM scan ).

Ah, If that's true that it's a cache, can we get rid of it? It seems like a premature optimization for a function that's only called basically during initialization

File: example/pages/index.tsx (hunk 1 above)
Comment (danieljackins) [reply example/pages/index.tsx]: love it!

File: src/__tests__/integration.test.ts [hunk 6]
```diff
@@ -175,6 +176,26 @@ describe('Initialization', () => {
...
+            ...xt,
+            load: async () => {
+              expect(window.analytics).toBeUndefined()
+              expect(getCDN()).toContain(overriddenCDNUrl)
+            },
+          },
+        ],
+      })
+      expect(fetchCalls[0][0]).toContain(overriddenCDNUrl)
+      expect.assertions(3)
```
Comment (chrisradek) [inline src/__tests__/integration.test.ts:196]: That's a neat trick :smile:

**Review by chrisradek — APPROVED** [review]
Looks great!

With regards to your question on if we need to setCdn for standalone/browser.umd, I think @pooyaj's explanation makes sense. If we can guarantee that we would only do the script tag lookup once with the changes you've made we can likely remove it - but I think to keep the scope of this PR focused I'd rather create a new issue or JIRA to track making that change/testing it.
